pkg/kubectl/cmd/debug.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"net/url"
+
+	"github.com/spf13/cobra"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	remotecommandconsts "k8s.io/apimachinery/pkg/util/remotecommand"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/kubernetes/pkg/api"
+	coreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/util/i18n"
+)
+
+var (
+	debug_example = templates.Examples(`
+		# Switch to raw terminal mode, sends stdin to 'bash' in a new container in pod 123456-7890
+		# named 'debug' from image 'debian and sends stdout/stderr from 'bash' back to the client
+		kubectl debug 123456-7890 -c debug -m debian -i -t -- bash -il`)
+)
+
+const (
+	debugUsageStr             = "expected 'debug POD_NAME [--] [COMMAND [ARG1 ... ARGN]]'.\nPOD_NAME is a required argument for the debug command.\nCOMMAND may be required if the container does not provide an ENTRYPOINT"
+	debugDefaultContainerName = "debug"
+	debugDefaultImageName     = "debian"
+)
+
+func NewCmdDebug(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer) *cobra.Command {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DebugContainers) {
```

Comment:
Having kubectl depend on Kubernetes repo is an anti pattern.  Don't do this.  It either needs to go in one of the staging repos (client-go, etc) or be copied into the kubectl packages